src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs

```diff
@@ -55,8 +87,14 @@ public override bool IsEvaluatableMethodCall(MethodCallExpression node)
 			var attributes = node.Method
 				.GetCustomAttributes(typeof(LinqExtensionMethodAttributeBase), false)
 				.Cast<LinqExtensionMethodAttributeBase>().ToArray();
-			return attributes.Length == 0 ||
-				attributes.Any(a => a.PreEvaluation == LinqExtensionPreEvaluation.AllowPreEvaluation);
+			if (attributes.Length > 0)
+				return attributes.Any(a => a.PreEvaluation == LinqExtensionPreEvaluation.AllowPreEvaluation);
+
+			if (_sessionFactory == null || _sessionFactory.Settings.LinqToHqlLegacyPreEvaluation ||
+				!_sessionFactory.Settings.LinqToHqlGeneratorsRegistry.TryGetGenerator(node.Method, out var generator))
+				return true;
```

Comment:
This also triggers a lookup of `IRuntimeMethodHqlGenerator` generators: finding them is a O(n) loop already done for each method lacking a regular generator. With this change, it is now done twice per such method.